### PR TITLE
Boopickle encoders

### PR DESCRIPTION
**Pull request by cquiroz**
We use [boopickle](https://boopickle.suzaku.io/) on our application with the encoders in this PR. I thought this could be generally useful in http4s though I'm not sure how widely used `boopickle` is.

This is also an opportunity, to receive any criticism you may have to my implementation. In particular I'm not sure if there is enough testing

I couldn't find an aggregate on `build.sbt` to include this into the overall build. Is it needed?

### CODE REVIEW

File: boopickle/src/main/scala/org/http4s/booPickle/BooPickleInstances.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      } else {
+        DecodeResult.failure[F, A](MalformedMessageBodyFailure("Invalid binary: empty body", None))
+      }
+    }
+
+  /**
+    * Create an `EntityDecoder` for `A` given a `Pickler[A]`
+    */
+  def booOf[F[_]: Sync, A: Pickler]: EntityDecoder[F, A] =
+    EntityDecoder.decodeBy(MediaType.`application/octet-stream`)(booDecoderByteBuffer[F, A])
```
Comment (rossabaker) [inline boopickle/src/main/scala/org/http4s/booPickle/BooPickleInstances.scala]: This is going to be hard to use, like `jsonOf` became hard to use in 0.18, isn't it?

File: boopickle/src/test/scala/org/http4s/booPickle/BoopickleSpec.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  "booOf" should {
+    "decode a class from a boopickle decoder" in {
+      val result = booOf[IO, Fruit]
+        .decode(Request[IO]().withBody(Banana(10.0): Fruit).unsafeRunSync, strict = true)
+      result.value.unsafeRunSync must_== Right(Banana(10.0))
+    }
+  }
+
+  checkAll("EntityCodec[IO, Json]", EntityCodecTests[IO, Fruit].entityCodec)
```
Comment (rossabaker) [inline boopickle/src/test/scala/org/http4s/booPickle/BoopickleSpec.scala]: Label is wrong on this test

File: build.sbt [hunk 3]
```diff
@@ -175,6 +175,15 @@ lazy val argonaut = libraryProject("argonaut")
...
   .dependsOn(core, testing % "test->test", jawn % "compile;test->test")
 
+lazy val boopickle = libraryProject("boopickle")
+  .settings(
+    description := "Provides Boopickle codecs for http4s",
+    libraryDependencies ++= Seq(
+      Http4sPlugin.boopickle
+    )
+  )
+  .dependsOn(core, testing % "test->test", jawn % "compile;test->test")
```
Comment (rossabaker) [inline build.sbt]: Why does this module depend on jawn?

File: boopickle/src/main/scala/org/http4s/booPickle/BooPickleInstances.scala (hunk 1 above)
Comment (cquiroz) [reply boopickle/src/main/scala/org/http4s/booPickle/BooPickleInstances.scala]: I’m not sure what you mean, could you elaborate?

Comment (cquiroz) [reply boopickle/src/main/scala/org/http4s/booPickle/BooPickleInstances.scala]: I see. I noted `decodeJson` was on `Message` but I assumed it was grandfathered. OTOH I don't think you want to keep adding `decodeXXX` methods. Or perhaps they  can be extension methods in the bundle?

**Comment by rossabaker** [comment]
New modules need to opt out of MiMa checking, because there isn't a prior version to check against.  I think there is an example of this in the Netty PR.

File: build.sbt [hunk 4]
```diff
@@ -176,6 +176,16 @@ lazy val argonaut = libraryProject("argonaut")
   )
   .dependsOn(core, testing % "test->test", jawn % "compile;test->test")
 
+lazy val boopickle = libraryProject("boopickle")
+  .settings(
+    description := "Provides Boopickle codecs for http4s",
+    libraryDependencies ++= Seq(
+      Http4sPlugin.boopickle
+    ),
+    mimaPreviousArtifacts := Set.empty
```
Comment (rossabaker) [inline build.sbt]: Just remember to remove this after we release 0.18.11.

**Pull request merged**